mvrx/src/main/kotlin/com/airbnb/mvrx/PersistState.kt

```diff
@@ -73,49 +72,71 @@ private fun assertCollectionPersistability(value: Any?) {
     }
     return this
 }
 
 /**
  * Updates the initial state object given state persisted with [PersistState] in a [Bundle].
  */
-internal fun <T : Any> Bundle.restorePersistedState(initialState: T): T {
+internal fun <T : MvRxState> Bundle.restorePersistedState(initialState: T): T {
+    val jvmClass = initialState::class.java
+    val constructor = jvmClass.constructors.firstOrNull { it.parameters.any { it.isAnnotationPresent(PersistState::class.java) } } ?: return initialState
+
     // If we don't set the correct class loader, when the bundle is restored in a new process, it will have the system class loader which
-    // can't unmarshall any custom classes.
-    val stateClass = initialState::class
-    classLoader = stateClass.java.classLoader
-    val constructor = stateClass.primaryConstructor ?: throw IllegalStateException("${stateClass.simpleName} has no primary constructor!")
-    val persistedConstructorParamNames = constructor.parameters.asSequence()
-        .filter { it.name != null }
-        .filter { it.annotations.any { it.annotationClass == PersistState::class } }
-        .map { it.name }
-        .toSet()
-    if (persistedConstructorParamNames.isEmpty()) {
-        return initialState
-    }
+    // can't unmarshal any custom classes.
+    classLoader = jvmClass.classLoader
+
+    // For data classes, Kotlin generates a static functionc called copy$default.
+    // The first parameter is the object to copy from.
+    // The next parameters are all of parameters to copy (it's jvm bytecode/java so there are no optional parameters in the generated method).
+    // The next parameter(s) are a bitmask. Each parameter index corresponds to one bit in the int.
+    //     If the bitmask is 1 for a given parameter then the new object will have the original object's value and the parameter value will be ignored.
+    //     If the bitmask is 0 then it will use the value from the parameter.
+    //     There is 1 bitmask for every 32 parameters. If there are 48 parameters, there will be 2 bitmasks. Parameter 33 will be the first bit of the 2nd bitmask.
+    // The last parameter is ignored. It can be null.
+    val copyFunction = jvmClass.declaredMethods.first { it.name == "copy\$default" }
+    val fieldCount = constructor.parameterCount
 
-    val copyMethod = stateClass.copyMethod()
-    val copyArgs = copyMethod.parameters.asSequence()
-        .filter { persistedConstructorParamNames.contains(it.name) }
-        .fold(mutableMapOf<KParameter, Any?>()) { map, param ->
-            // We do the containsKey check to differentiate between a missing key and one that is explicitly null.
-            if (containsKey(param.name)) map[param] = this[param.name]
-            map
+    // There is 1 bitmask for each block of 32 parameters.
+    val parameterBitMasks = IntArray(fieldCount / 32 + 1) { 0 }
+    val parameters = arrayOfNulls<Any?>(fieldCount)
+    parameters[0] = initialState
+    for (i in 0..(fieldCount - 1)) {
```

Comment:
iterate  from 1? the first parameter is the initial state

Reply:
@elihart The first parameter of the copy function is the initial state but this is iterating over the constructor parameters